### ISSUE TITLE
Hooks API: support events with multiple signature args

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#594]
When we were designing the hooks API, we discussed the concept of signature args for events at great length. The ultimate conclusion of this discussion was in WordPoints/hooks-api#78, where it was concluded that:

> [E]vents have at most only a single "primary" arg.

However, in https://github.com/WordPoints/buddypress/issues/19#issuecomment-268369923 we discovered that we hadn't considered one particular class of events, those in which a relationship is forged between two args. Such an event can be reversed when the relationship is later broken, so it isn't a repeatable event. We noted (https://github.com/WordPoints/buddypress/issues/19#issuecomment-268375378):

>Currently, if we don't incorporate both args into the signature of the event, the reversals extension will not be able to detect reversals correctly (the legacy points reversals and repeat blocker extensions would also suffer, but they aren't really relevant here). What would happen, if we made these both stateful args, would be that when one hit was reversed, all hits would be, for all users. Which is why reversible events need to have signature args. If we just register one of the args, it will still cause problems: if we register only the user, hits for all favorites for that user will be reversed when one favorite is removed; if we register only the activity, hits for all users for that activity we be reversed when one user un-favorites it. So we really do have to have two signature args here for this to work as we want. 

IIRC, I did consider something somewhat similar to this situation, but I thought of it only in terms of the event targets, not the reversals extension. As far as the targets go, in theory both args don't need to be supplied, since the args are related, and they don't need to both be signature anyway.

**Comment by JDGrimes** [comment]
## Entity args object

This means we need to replace the `get_primary_arg()` method on the event args hierarchy object with `get_signature_args()`. Previously, we just saved the slug of the first non-stateful arg as the primary arg. But now I'm thinking it might be easier to just save the entity objects in a property themselves, rather than just the slugs. The only downside to that is that it means we need to consider what to do when `add_entity()` and `remove_entity()` are called. Do we need to remove the signature args in the `remove_entity()` method? I suppose so, although I would note that we don't really use this method anywhere right now. As far as `add_entity()` goes, previously we just assumed that all entities added were stateful, we didn't set the primary arg if a non-stateful entity was added. ~~But I guess really we ought to check if the entity is stateful~~ Entities can't be checked for that, because they can't be stateful or non-stateful, only args can. So if signature args are to be added, they would have to be added via a different method. But I think that basically we concluded that adding signature args would be changing the very nature of the event, so they should really be present here right from the start. The main reason those methods exist is just to provide complete handling for the parent hierarchy class anyway, not specifically for the event args handling.

**Comment by JDGrimes** [comment]
The same really goes for removing signature args too, I guess: doing so changes the very nature of the event. So maybe we should just not worry about handling that, since anyone doing so is probably `_doing_it_wrong()`. Perhaps we should really even consider not allowing signature args to be removed at all, although this behavior is different from before. ~~We currently would let primary args be removed, and would simply return `false` if the primary arg was not found.~~ Actually, we currently don't provide any grace for the removal of the primary arg. We attempt to access it even though if it was removed it wouldn't be in the array. So not handling removal would actually be consistent with what we already do.

**Comment by JDGrimes** [comment]
This `get_primary_arg()` method is used in three locations:

- `wordpoints_hooks_get_event_primary_arg_guid_json()`
- The legacy points repeat blocker extension.
- The legacy points reversals extension.

**Comment by JDGrimes** [comment]
## Get GUID JSON function

For that first one, we'll need to replace it with a new function, `wordpoints_hooks_get_event_signature_arg_guids_json()`, that can handle multiple signature args. I'm not sure how best to handle the JSON for them though. Should we use a keyed or unkeyed array? We could key the array by the arg slug, which would enable us to ensure that it always had the same sort-order more easily (although we could also do that in theory just using the GUIDs). Really though, it wouldn't provide necessary information, because the GUIDs already tell what kind of entity they are. However, I suppose two signature args could be of the same entity type, in which case we wouldn't know which GUID was for which. So I suppose we do need to key the array by GUID.

**Comment by JDGrimes** [comment]
This function is used only in the fire object, in `log_hit()` and `get_matching_hits_query()`, to generate the value for the `primary_arg_guid` column in the query. Note that this means that the new function will need to continue to return just the GUID for the arg if there is only one, rather than wrapping it in an array. Otherwise the queries will not match up, since different strings would be produced, and everything would break.

**Comment by JDGrimes** [comment]
This also makes me wonder whether we should consider updating the name of that column. It should really be `signature_arg_guids` now. But I'm not sure if it is worth it to make the change, although it might be confusing to leave it. I guess we can revisit that a bit later.

**Comment by JDGrimes** [comment]
## Legacy extensions

In the legacy repeat blocker extension we get the primary arg just so that we can search for the entity ID in the points logs meta. (The same for the legacy reversal extension.) In this case I see no reason to worry about multiple signature args, because we never had more than one signature arg in the legacy hooks. At least, we didn't in core, although I suppose that somebody else could have. However, the improbability of that dictates that we not worry about handling multiple signature args here. We'll just always assume that the first signature arg is the one we're interested in, since there should always be only one for events using this extension anyway. If we do find we need to handle multiple signature args here in the future, we can always add that.

**Comment by JDGrimes** [comment]
## Changing column name

We could change the column name, and I'd like to, but we need to consider backward compatibility. There is no reason that anybody should be referencing that column directly in their own queries though, since we provide a query class for this table. And so we can provide backward compatibility in the query class, while changing the underlying table itself, I think.

**Comment by JDGrimes** [comment]
For some reason our tests are getting several failures on Travis that aren't causing issues locally:

```
There were 2 errors:

1) WordPoints_Un_Installer_Base_Test::test_uninstall_table

WordPress Database Error: Table 'wptests_test' already exists [CREATE TEMPORARY TABLE `wptests_test` ( `id` BIGINT );]

/home/travis/build/WordPoints/wordpoints/dev-lib/phpunit/classes/error/handler/database.php:51
/tmp/wordpress/src/wp-includes/wp-db.php:1360
/tmp/wordpress/src/wp-includes/wp-db.php:1816
/home/travis/build/WordPoints/wordpoints/tests/phpunit/tests/un-installer-base.php:2957

2) WordPoints_Un_Installer_Base_Test::test_uninstall__table

WordPress Database Error: Table 'wptests_test' already exists [CREATE TEMPORARY TABLE `wptests_test` ( `id` BIGINT );]

/home/travis/build/WordPoints/wordpoints/dev-lib/phpunit/classes/error/handler/database.php:51
/tmp/wordpress/src/wp-includes/wp-db.php:1360
/tmp/wordpress/src/wp-includes/wp-db.php:1816
/home/travis/build/WordPoints/wordpoints/tests/phpunit/tests/un-installer-base.php:2996
```

**Comment by JDGrimes** [comment]
Apparently it isn't an error arising from anything we did here, I just [reran a Travis job on `master`](https://travis-ci.org/WordPoints/wordpoints/jobs/183102760), and it had the same issues. So I guess it must have something to do with MySQL version changes or something like that.

**Comment by JDGrimes** [comment]
The tests are failing when WordPoints is installed per-site on multisite:

```
1) WordPoints_2_3_0_Update_Test::test_hooks_hit_table_primary_arg_guid_column_renamed

Failed asserting that null matches expected 'signature_arg_guids'.

/home/travis/build/WordPoints/wordpoints/tests/phpunit/tests/update/2-3-0.php:85
```

The cause seems to be that when we check the version, `$this->get_db_version()` is returning false:

```php
		if ( version_compare( $this->get_db_version(), '2.1.0-alpha-3', '<' ) ) {
			return;
		}
```

This causes the version to be evaluated as less than 2.1.0-alpha-3, and thus updating the table is skipped.

The reason that `get_db_version()` returns false is because it pulls the version from the site options based on the `context`, which in this case is `'network'`, since this is a network-wide table and so is being updated in the network-wide method. However, WordPoints is not installed network-wide, so the version is not saved in the site-wide options table, but in the per-site option tables.

We need to either take that into account here, or more probably, fix the way that `get_db_version()` operates.

**Comment by JDGrimes** [comment]
Fixed in f43224f.
